Discrete colormap/colorbar option

jhamman opened the pull request:
This PR adds a discrete colormap / colorbar option to xray plots.

closes #500 

```python
import matplotlib.pyplot as plt
import numpy as np
import xray

# sample DataArray
x = np.arange(start=0, stop=10, step=2)
y = np.arange(start=9, stop=-7, step=-3)
xy = np.dstack(np.meshgrid(x, y))
distance = np.linalg.norm(xy, axis=2)
distance = xray.DataArray(distance, list(zip(('y', 'x'), (y, x))))

# Sample plots
fig, axes = plt.subplots(2, 2, figsize=(8, 8), sharex=True, sharey=True)
distance.plot(ax=axes[0, 0])
axes[0, 0].set_title('Default')
distance.plot(vmin=0, vmax=12, levels=6, ax=axes[1, 0])
axes[1, 0].set_title('Setting with integer levels')
distance.plot(levels=[1, 2, 4, 5, 7, 9], extend='both', cmap='Spectral', ax=axes[0, 1])
axes[0, 1].set_title('Setting with list of levels')
flatui = ["#9b59b6", "#3498db", "#95a5a6", "#e74c3c", "#34495e", "#2ecc71"]
distance.plot(levels=[1, 2, 4, 5, 7], cmap=flatui, extend='both', ax=axes[1, 1])
axes[1, 1].set_title('Using custom list of colors')
```
Produces these plots:

![discrete_example](https://cloud.githubusercontent.com/assets/2443309/9027555/93d8220c-390e-11e5-8184-9f1b75b9ebfe.png)

There are potential issues that I'd still like to look into a bit further.

1.  Does discretizing `cmap` impact the `contour` and `contourf` plot types? We may need some logic to only discretize when the plot type is `imshow` or `pcolormesh`.
2.  It is possible, in some cases, to end up masking out values equal to `vmax`:
    ```python
    distance.plot(levels=6)
    ```
    Produces this plot:

    ![missing_corner](https://cloud.githubusercontent.com/assets/2443309/9027585/0e5ea5f4-3910-11e5-911a-7be312a59876.png)

    In the first example, we corrected for this by setting `extend='max'`.

jhamman commented on xray/plot/plot.py:
```diff
@@ -257,7 +260,55 @@ def _determine_cmap_params(plot_data, vmin=None, vmax=None, cmap=None,
+
+    try:
+        from seaborn.apionly import color_palette
+        pal = color_palette(cmap, n_colors=n_colors)
```
This section here could use a bit more thought.  The nice part about using seaborn's `color_palette` is that it allows us to access all its color schemes. 

There are two downsides:

1.  We add seaborn as a dependency or hope to recover correctly (As shown below)
2.  `color_palette` doesn't work with `LinearSegmentedColormap` objects which is how `viridis` is currently stored.

shoyer replied:
This actually looks like a pretty reasonable solution to me. It basically works the exact same way -- the only difference is that if seaborn is installed you can also use seaborn palette names.

For viridis, we should probably switch from using `LinearSegmentedColormap` to using `ListedColormap` instead -- that's how the final matplotlib PR ended up: https://github.com/matplotlib/matplotlib/pull/4707/files

shoyer commented on xray/plot/plot.py:
```diff
@@ -257,7 +260,55 @@ def _determine_cmap_params(plot_data, vmin=None, vmax=None, cmap=None,
+        ext_n = 0
+
+    if isinstance(levels, int):
+        cticks = np.linspace(vmin, vmax, num=levels + 1, endpoint=True)
```
I wonder if we can deal with the non-colored endpoints issue by simply adding some small epsilon to the last tick-point?

jhamman commented:
These last two commits have resolved the two issues I brought up above.

shoyer commented:
Some feedback:

1. matplotlib seems to have nice heuristics for picking level splits with contour plots. I wonder if it would be feasible to copy that logic here when setting our own discrete levels? Compare:
`ds.t2m[3].plot(levels=6)`
![image](https://cloud.githubusercontent.com/assets/1217238/9033599/8d201d7e-397d-11e5-9a78-2e6bdb55725a.png)
vs `ds.t2m[3].plot.contourf()`
![image](https://cloud.githubusercontent.com/assets/1217238/9033607/9b115bfa-397d-11e5-8978-25626ec1419a.png)

2. The logic for inferring the `extend` keyword argument needs to be updated to handle `levels`:
`ds.t2m[3].plot(robust=True, levels=[200, 250, 260, 270, 280, 290, 350])`
![image](https://cloud.githubusercontent.com/assets/1217238/9033635/de3e2944-397d-11e5-86a8-a88c57f6ae4c.png)

3.`imshow`, `pcolormesh` and `contourf` seem to deal with an explicit list of `levels` the exact same way. That's a relief. The plots look exactly the same.

4. As you noted, `contourf` currently crashes if you supply `levels` as an integer. This should be fixed, probably by converting `levels` into a 1d array / list if they are supplied as an integer before passing it into `contourf`. This would actually be more broadly useful, because currently there is no way to supply a non-default integer number of levels with `contourf` -- matplotlb appears to only expose the `N` argument when it is used positionally.

shoyer commented:
Looking at the matplotlib source code, here's how they figure out levels in contour/contourf:
https://github.com/matplotlib/matplotlib/blob/714d18788320325d0bff75184f62d472f67ceb91/lib/matplotlib/contour.py#L1143

It looks like they use `MaxNLocator` to do this:
http://matplotlib.org/api/ticker_api.html#matplotlib.ticker.MaxNLocator

It should be pretty straightforward to use that here.

jhamman commented:
> The logic for inferring the extend keyword argument needs to be updated to handle levels:

How do you want to handle this?  Options for when `levels` is provided and `extend == None`:

1. Don't extend.  This will result in data outside the range of `levels` being masked out, 
1. Extend if the min/max of the data exceed the range of `levels`, or 
1. Require extend to be specified when `levels` is provided.

I will add the `MaxNLocator` ticker for integer `levels`.  I agree this is the way to go. 

I think this logic will also help address your 4th point: 
> ...contourf currently crashes if you supply levels as an integer. This should be fixed, probably by converting levels into a 1d array / list if they are supplied as an integer before passing it into contourf. This would actually be more broadly useful, because currently there is no way to supply a non-default integer number of levels with contourf -- matplotlb appears to only expose the N argument when it is used positionally.

shoyer commented:
> Options for when levels is provided and extend == None:

I would opt for your solution 2. I think this basically means using `levels[0]` and `levels[-1]` if levels are provided explicitly instead of `vmin`/`vmax` (which are being ignored anyways).

jhamman commented:
@shoyer  - This last commit (119c1c8) should have addressed the issues you brought up. I updated my examples a bit to show how things are working:

```python
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import xray

# sample DataArray
x = np.arange(start=0, stop=10, step=2)
y = np.arange(start=9, stop=-7, step=-3)
xy = np.dstack(np.meshgrid(x, y))
distance = np.linalg.norm(xy, axis=2)
distance = xray.DataArray(distance, list(zip(('y', 'x'), (y, x))))

# Sample plots
fig, axes = plt.subplots(3, 2, figsize=(8, 8), sharex=True, sharey=True)
distance.plot(ax=axes[0, 0])
axes[0, 0].set_title('Default')
distance.plot(vmin=0, vmax=12, levels=6, ax=axes[1, 0])
axes[1, 0].set_title('Setting with integer levels')
distance.plot(levels=[1, 2, 4, 5, 7, 9], extend='max', cmap='Spectral', ax=axes[0, 1])
axes[0, 1].set_title('Setting with list of levels')
flatui = ["#9b59b6", "#3498db", "#95a5a6", "#e74c3c", "#34495e", "#2ecc71"]
distance.plot(levels=[1, 2, 4, 5, 7], cmap=flatui, extend='both', ax=axes[1, 1])
axes[1, 1].set_title('Using custom list of colors')
distance.plot.contourf(ax=axes[2, 0])
axes[2, 0].set_title('Default contourf')

distance.plot.contourf(levels=4, ax=axes[2, 1])
axes[2, 1].set_title('contourf w/ int levels')

plt.tight_layout()
```

If the user sets the `extend`, `vmax`, or `vmin` arguments, those are not overridden.  This is shown in the top right plot where I set `extend='max'` even though the min of the data is less than the min of `levels`.  This results in the data less than the min of `levels` not being show. For the same reason, we also don't use the `MaxNLocator` if `vmin` or `vmax` was provided.

![discrete_example](https://cloud.githubusercontent.com/assets/2443309/9053006/561c6278-3a23-11e5-9627-f4273eb0e759.png)

shoyer commented:
Decided to try reorganizing things and got sucked in, so I ended up making a PR to your fork with some tweaks:
https://github.com/jhamman/xray/pull/1

This assuredly still needs more tests -- mostly to verify that every plot types ends up with consistent levels.

Here's my demo notebook:
https://gist.github.com/shoyer/e6fafbc672021600d9f3

and a teaser:
![image](https://cloud.githubusercontent.com/assets/1217238/9055420/d61ca21e-3a3f-11e5-8d7b-172545658c34.png)

The main difference in functionality is in my PR is that `vmin` and `vmax` *cannot* be set independently of `levels`, if levels is provided as a list. That seemed a little too flexible, and this way we can get a consistent result across all four plot types. Are there any cases where you *would* want to set `vmin`/`vmax` to something other than the first and last levels?

Also, I entirely replaced the levels calculation in contour/contourf with our calculated levels instead of merely trying to replicate it. That ensures things are guaranteed to be exactly identical across all plot types.

jhamman commented:
This looks good.  Thanks for doing that.

>Are there any cases where you would want to set vmin/vmax to something other than the first and last levels?

Yes, but they are rare and you could pass in a custom `norm` in those cases. 

>This assuredly still needs more tests -- mostly to verify that every plot types ends up with consistent levels.

I'll think a bit more about how to test these aspects.

----

There are three things left to do here (this has gotten a big bigger than I was expecting):

- [ ] Decide what to do with seaborn's handling of `jet`
- [ ] Fill in tests
- [ ] Add an example in the plot docs

jhamman commented:
@shoyer - I added a bunch of tests this morning and checked off the list of documentation and error handling items.  Take another look and let me know if you think we need anymore tests or changes to the docs.

shoyer commented on doc/plotting.rst:
```diff
@@ -201,7 +201,7 @@ distance from a 2d grid point to the origin.
     distance = np.linalg.norm(xy, axis=2)
 
-    distance = xray.DataArray(distance, zip(('y', 'x'), (y, x)))
+    distance = xray.DataArray(distance, zip(('y', 'x'), (y, x))))
```
the extra parenthesis breaks the doc build

shoyer commented:
These tests look great -- thanks!

shoyer commented:
@clarkfitzg any thoughts here?

clarkfitzg commented on xray/plot/plot.py:
```diff
@@ -242,22 +246,87 @@ def _determine_cmap_params(plot_data, vmin=None, vmax=None, cmap=None,
+    """
+    import matplotlib as mpl
+
+    def extension_colors(extend):
```
Probably could get away without the nested function here?

clarkfitzg commented on xray/plot/plot.py:
```diff
@@ -337,7 +411,8 @@ def _plot2d(plotfunc):
     def newplotfunc(darray, ax=None, xincrease=None, yincrease=None,
                     add_colorbar=True, vmin=None, vmax=None, cmap=None,
-                    center=None, robust=False, extend=None, **kwargs):
+                    center=None, robust=False, extend=None, levels=None,
```
We'll want to update the signature for `plotmethod` also to include this. My fault on having to maintain two signatures like this- have been thinking if there's a better way.

jhamman replied:
I went ahead and added `levels` to the signature.

clarkfitzg commented on xray/test/test_plot.py:
```diff
@@ -165,18 +167,136 @@ def test_plot_nans(self):
+
+        orig_levels = [0, 1, 2, 3, 4, 5]
+        # vmin and vmax should be ignored if levels are explicitly provided
+        vmin, vmax, cmap, extend, levels, cnorm = _determine_cmap_params(
```
Another possible implementation to consider is having `_determine_cmap_params` return `locals()`. Then we can use a dictionary rather than doing tuple unpacking many times. Labeled data FTW :)

jhamman replied:
I don't have strong feelings about this either way.  @shoyer, thoughts?

shoyer replied:
I think we'll need to filter locals() if we take that approach to avoid intermediate variables. But you could construct a dict manually for return value and use that to update kwargs. We are definitely getting to the point of too many positional arguments.

clarkfitzg commented on xray/test/test_plot.py:
```diff
@@ -165,18 +167,136 @@ def test_plot_nans(self):
+                               ('neither', [0, 5, 10, 15]),
+                               ('min', [2, 5, 10, 15])]:
+            for kind in ['imshow', 'pcolormesh', 'contourf', 'contour']:
+                primative = getattr(self.darray.plot, kind)(levels=levels)
```
spelling - primitive

clarkfitzg commented on xray/test/test_plot.py:
```diff
@@ -284,6 +404,13 @@ def test_extend(self):
         artist = self.plotmethod(vmin=-10, vmax=0)
         self.assertEqual(artist.extend, 'max')
 
+    def test_levels(self):
```
Not exactly sure what this does since the `plotmethod` doesn't seem to have a levels arg at the moment.

jhamman replied:
I'm fairly sure it was passed through as a `kwarg`. I added it to the signature so its there now.

clarkfitzg commented:
Nice work @jhamman @shoyer. Tests look pretty thorough and I like all the helper functions.

shoyer commented:
@jhamman you should merge this once you and @clarkfitzg are happy. I'm about go offline for 6 days (backpacking in the Olympics).

jhamman commented:
@shoyer - have a good time.

@clarkfitzg - I think I've addressed all your comments, including returning a `dict` from `_determine_cmap_params`.  If there is nothing else, I'll merge this tomorrow.

clarkfitzg commented:
Yes, looks good. Merge away!

clarkfitzg commented on doc/conf.py:
```diff
@@ -42,6 +42,11 @@
 except ImportError:
     print "no ipython"
+try:
+    import seaborn
```
Docs look good on the server- http://xray.readthedocs.org/en/latest/plotting.html#discrete-colormaps, but for some reason this line causes my local build to to hang on `reading sources... [ 98%] plotting`. 

Also the first few plots show up in a window during the build, like in matplotlib interactive mode. Not sure why this happens- maybe caused by some config files somewhere on my machine.

jhamman replied:
That is really odd. However, I just built the docs and was not able to reproduce your described behavior.

What happens if you put `seaborn.reset_orig()`  after the import.

shoyer replied:
I also had an issue where building the docs on this branch popped up matplotlib windows which paused the build on my laptop. Not quite sure what's going on -- some import may be switching the matplotlib backend.

On Thu, Aug 6, 2015 at 11:24 AM, Joe Hamman <notifications@github.com>
wrote:

>> @@ -42,6 +42,11 @@
>>      print "ipython: %s, %s" % (IPython.__version__, IPython.__file__)
>>  except ImportError:
>>      print "no ipython"
>> +try:
>> +    import seaborn
> That is really odd. However, I just built the docs and was not able to reproduce your described behavior.
> What happens if you put `seaborn.reset_orig()`  after the import.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/xray/xray/pull/509/files#r36447826

clarkfitzg replied:
Happens when I run `make clean` and then `make html`. Ran into something like this earlier and fixed it by using `matplotlib.use('Agg')`. PR coming along shortly.
